File: Source/Android/luascriptcore/src/main/java/cn/vimfung/luascriptcore/LuaFunction.java
package cn.vimfung.luascriptcore;

/**
 * Lua方法对象
 * Created by vimfung on 16/11/1.
 */
public class LuaFunction extends LuaBaseObject
{
    private LuaContext _context;

    /**
     * 创建方法对象
     * @param nativeId  本地对象标识
     * @param context 上下文对象
     */
    protected LuaFunction(int nativeId, LuaContext context)
    {
        super(nativeId);
        _context = context;
    }

    /**
     * 调用方法
     * @param arguments 参数列表
     * @return  返回值
     */
    public LuaValue invoke(LuaValue[] arguments)
    {
        return invoke(arguments, null);
    }

    public LuaValue invoke(LuaValue[] arguments, LuaScriptController scriptController)
    {
        return LuaNativeUtil.invokeFunction(_context, this, arguments, scriptController);
    }

    /**
     * 获取上下文对象
     * @return 上下文对象
     */
    public LuaContext getContext()
    {
        return _context;
    }
}

File: Source/Unity3D/UnityProject/Assets/Plugins/LuaScriptCore/LuaFunction.cs
using System;
using System.Collections.Generic;
using System.Runtime.InteropServices;

namespace cn.vimfung.luascriptcore
{
	/// <summary>
	/// Lua方法
	/// </summary>
	public class LuaFunction : LuaBaseObject
	{
		private LuaContext _context;

		/// <summary>
		/// 初始化
		/// </summary>
		/// <param name="decoder">对象解码器</param>
		public LuaFunction (LuaObjectDecoder decoder)
			: base(decoder)
		{
			int contextId = decoder.readInt32 ();
			_context = LuaContext.getContext (contextId);
			luaObjectId = decoder.readString ();
		}

		/// <summary>
		/// 序列化对象
		/// </summary>
		/// <param name="encoder">对象编码器.</param>
		public override void serialization(LuaObjectEncoder encoder)
		{
			base.serialization (encoder);

			encoder.writeInt32 (_context.objectId);
			encoder.writeString (luaObjectId);
		}

		/// <summary>
		/// 调用
		/// </summary>
		/// <param name="arguments">Arguments.</param>
		public LuaValue invoke(List<LuaValue> arguments)
		{
			return invoke (arguments, null);
		}

		/// <summary>
		/// 调用方法
		/// </summary>
		/// <param name="arguments">参数列表</param>
		/// <param name="scriptController">脚本控制器</param>
		public LuaValue invoke(List<LuaValue> arguments, LuaScriptController scriptController)
		{
			int scriptControllerId = 0;
			if (scriptController != null)
			{
				scriptControllerId = scriptController.objectId;
			}

			IntPtr funcPtr = IntPtr.Zero;
			IntPtr argsPtr = IntPtr.Zero;
			IntPtr resultPtr = IntPtr.Zero;

			LuaObjectEncoder funcEncoder = new LuaObjectEncoder (_context);
			funcEncoder.writeObject (this);

			byte[] bytes = funcEncoder.bytes;
			funcPtr = Marshal.AllocHGlobal (bytes.Length);
			Marshal.Copy (bytes, 0, funcPtr, bytes.Length);

			if (arguments != null)
			{
				LuaObjectEncoder argEncoder = new LuaObjectEncoder (_context);
				argEncoder.writeInt32 (arguments.Count);
				foreach (LuaValue value in arguments)
				{
					argEncoder.writeObject (value);
				}

				bytes = argEncoder.bytes;
				argsPtr = Marshal.AllocHGlobal (bytes.Length);
				Marshal.Copy (bytes, 0, argsPtr, bytes.Length);
			}

			int size = NativeUtils.invokeLuaFunction (_context.objectId, funcPtr, argsPtr, scriptControllerId, out resultPtr);

			if (argsPtr != IntPtr.Zero)
			{
				Marshal.FreeHGlobal (argsPtr);
			}
			if (funcPtr != IntPtr.Zero)
			{
				Marshal.FreeHGlobal (funcPtr);
			}

			if (size > 0)
			{
				return LuaObjectDecoder.DecodeObject (resultPtr, size, _context) as LuaValue;
			}

			return new LuaValue ();
		}

		/// <summary>
		/// 获取上下文对象
		/// </summary>
		/// <value>上下文对象.</value>
		public LuaContext context
		{
			get
			{
				return _context;
			}
		}
	}
}

